### PR TITLE
Air Boundary - Part 2 - Cleanup and Fix Daylighting

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
Followup to #7455 "New Air Thermal Boundary Construction Type"
 - [x] Address comments and cleanup
 - [x] Fix Daylighting with GroupedZones option - *Getting illuminance values in the core zone (Space5-1) that are within 0.1 lux of this comparison run which combines the 3 grouped zones into a single zone. [5ZoneAirCooled_AirBoundaries_Daylighting](https://github.com/NREL/EnergyPlus/blob/AirWallCleanupAndDaylighting/testfiles/5ZoneAirCooled_AirBoundaries_Daylighting.idf) vs 
[5ZoneAirCooled_CombineZones.idf.txt](https://github.com/NREL/EnergyPlus/files/3537597/5ZoneAirCooled_CombineZones.idf.txt)*

 - [x] Disable daylighting and solar with InteriorWindow option - *Currently this is showing no daylighting or solar gains in Space5-1*
*This will have to get fixed in a Part 3 pull request.*

 - [x] Add engineering ref discussion of how the daylighting zones and enclosures are handled.
 - [x] Extend unit tests
Coverage of new air boundary functions:
`HeatBalanceManager::CreateAirBoundaryConstructions`
`HeatBalanceIntRadExchange::AlignInputViewFactors`
`HeatBalanceIntRadExchange::InitSolarViewFactors`
`SurfaceGeometry::SetupRadiantEnclosuresAndAirBoundaries`
`SurfaceGeometry::SetupSolarEnclosuresAndAirBoundaries`

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by rraustad** [comment]
Easier diff's to review! I see these 3 changes in the mtd file. Not sure if CI got confused on the last addition/deletion and carried over text color (showing a false diff) or if this is a real change. Just pointing to a diff.

These were added, blue color text but has no + beside it
Meters for 380,SB35:Surface Inside Face Solar Radiation Heat Gain Energy [J]
Meters for 379,SB35:Surface Inside Face Net Surface Thermal Radiation Heat Gain Energy [J]

This was deleted, green color test but no - beside it
Meters for 403,F4-1:Surface Inside Face Solar Radiation Heat Gain Energy [J]

**Comment by mjwitte** [comment]
@rraustad If you look at the err diffs, the new example file had some duplicate variables in the custom meter objects. Those are fixed here, so the errors are gone and the mtd changed. I'm working on some plots now to show that the daylighting results are (almost exactly) the same with combined zones vs grouped air boundary zones. Will post soon.

**Comment by mjwitte** [comment]
Here are plots of daylighting reference point illuminance and glare index for 01/01 and 07/01. 
Space145 is the combined zone run with develop. 
The other case, run with this branch, groups the same three zones but they remain distinct zones with air boundary surfaces.
Space 145 Ref point 1 is in the same place as Space1-1Ref Point 1.
Space 145 Ref point 2 is in the same place as Space5-1 Ref Point 1
The glare index is identical for both cases.
I expect the illuminances to be exactly the same also, but they differ by a small amount, <0.1 lux
Next, I need to test a case using the air boundary interior window option.

image

**Review by rraustad — APPROVED** [review]
Quit while your ahead.

**Comment by mjwitte** [comment]
But, just, one, more, thing . . .

**Comment by rraustad** [comment]
So the recent x86_64-Linux-Ubuntu-18.04-gcc-7.4 diff's now showing are similar to what I saw on different branches that have been merged. That CI is comparing to a new baseline and causing diff's. This branch likely has only the 1 AUD, 1 ERR and 1 MTD diff.

**Comment by mjwitte** [comment]
@Myoldmopar This is ready for review. I can address comments here or in a new PR when I return.

File: doc/engineering-reference/src/daylighting-and-window-calculations/daylighting-calculations.tex [hunk 1]
```diff
@@ -10,8 +10,14 @@ \section{Daylighting Calculations}\label{daylighting-calculations}
 \item
     \emph{Daylight factors}, which are ratios of interior illuminance or luminance to exterior horizontal illuminance, are calculated and stored.
     The user specifies the coordinates of one or more reference points in each daylit zone.
-    EnergyPlus then integrates over the area of each exterior window in the zone to obtain the contribution of direct light from the window to the illuminance at the reference points,
-    and the contribution of light that reflects from the walls, floor and ceiling before reaching the reference points.
+    EnergyPlus then integrates over the area of each exterior window in the zone (or enclosure) to obtain the contribution of direct light from the window to the illuminance at the reference points,
+    and the contribution of light that reflects from the walls, floor and ceiling before reaching the reference points. 
+    
+    If a zone has surfaces with air boundaries (Construction:AirBoundary) 
+    using the \textit{GroupedZones} method for solar and daylighting then all of the surfaces within the grouped enclosure are included in the "zone". Each daylighting control and reference point
```
Comment (Myoldmopar) [inline doc/engineering-reference/src/daylighting-and-window-calculations/daylighting-calculations.tex:17]: Ideally, quotes in LaTeX are `` (two backticks) and '' (two apostrophes), which allows LaTeX to render the actual character as it sees fit.  No big deal, just something to note for the future.

File: doc/input-output-reference/src/overview/group-surface-construction-elements.tex [hunk 2]
```diff
@@ -4082,37 +4082,28 @@ \subsubsection{Inputs}\label{inputs-38-01}
...
-  
-  \item
-  \textbf{InteriorWindow} - The air boundary will be modeled as a perfectly transmitting interior window. As with other interior windows, all direct solar passing through the interior window will be diffuse in the adjacent zone.
-\end{itemize}
+\begin{description}
+  \item[GroupedZones] The two zones separated by this air boundary will be grouped together into a combined zone for solar distribution and daylighting. If a given zone has an air boundary with more than one zone, then all of the connected zones will be grouped together. For example, if there is an air  boundary between zones A and B, and another air boundary between zones B and C, all three zones (A, B, and C) will be grouped into a single zone.
+  \item[InteriorWindow] The air boundary will be modeled as a perfectly transmitting interior window. As with other interior windows, all direct solar passing through the interior window will be diffuse in the adjacent zone.
+\end{description}
```
Comment (Myoldmopar) [inline doc/input-output-reference/src/overview/group-surface-construction-elements.tex:4088]: Yay for `{description}`s!

File: src/EnergyPlus/DataSurfaces.hh [hunk 3]
```diff
@@ -1085,9 +1084,8 @@ namespace DataSurfaces {
         Real64 VentingOpenFactorMultRep;        // Window/door opening modulation multiplier on venting open factor, for reporting
         Real64 InsideTempForVentingRep;         // Inside air temp used to control window/door venting, for reporting (C)
         Real64 VentingAvailabilityRep;          // Venting availability schedule value (0.0/1.0 = no venting allowed/not allowed)
-        Array1D<Real64> IllumFromWinAtRefPtRep; // Illuminance from window at reference point #1 [lux]
-        Array1D<Real64> LumWinFromRefPtRep;     // Window luminance as viewed from reference point #1 [cd/m2]
-        Real64 LumWinFromRefPt2Rep;             // Window luminance as viewed from reference point #2 [cd/m2]
```
Comment (Myoldmopar) [inline src/EnergyPlus/DataSurfaces.hh:1090]: Hmm. Unexpected to see a variable just go away, but I'm happy with that. :+1:

File: src/EnergyPlus/DataViewFactorInformation.hh [hunk 4]
```diff
@@ -76,14 +76,15 @@ namespace DataViewFactorInformation {
...
         Array1D<Real64> Azimuth;            // Azimuth angle of the surface (in degrees)
         Array1D<Real64> Tilt;               // Tilt angle of the surface (in degrees)
-        Array1D_int SurfacePtr;             // Surface ALLOCATABLE (to Surface derived type)
-        Real64 FloorArea;                   // Floor area of zone(s) in radiant enclosure
+        Array1D_int SurfacePtr;             // Surface number for surfaces in this enclosure
+        Real64 FloorArea;                   // Floor area of zone(s) in enclosure
         Real64 ExtWindowArea;               // Exterior window area
         Real64 TotalSurfArea;               // Total surface area
         Array1D<Real64> SolAbsorptance;     // Surface solar absorptance
+        int TotalEnclosureDaylRefPoints;    // Total number of daylighting reference points in enclosure
```
Comment (Myoldmopar) [inline src/EnergyPlus/DataViewFactorInformation.hh:84]: Reasonable addition.

File: src/EnergyPlus/DaylightingManager.cc [hunk 5]
```diff
@@ -345,27 +327,25 @@ namespace DaylightingManager {
...
         AInsTot = 0.0;
         // Sum of products of inside surface area * vis reflectance
         ARHTOT = 0.0;
         // Area sum and area * reflectance sum for different orientations
         AR = 0.0;
         ARH = 0.0;
-        // Loop over surfaces
-        for (ISurf = Zone(ZoneNum).SurfaceFirst; ISurf <= Zone(ZoneNum).SurfaceLast; ++ISurf) {
+        // Loop over surfaces in the zone's enclosure
+        auto & thisEnclosure(DataViewFactorInformation::ZoneSolarInfo(Zone(ZoneNum).SolarEnclosureNum));
```
Comment (Myoldmopar) [inline src/EnergyPlus/DaylightingManager.cc:340]: And so this won't double-count anything when this is called for the next zone in the same enclosure, right?

File: src/EnergyPlus/DaylightingManager.cc [hunk 6]
```diff
@@ -4620,6 +4579,61 @@ namespace DaylightingManager {
...
+                                                        "Zone",
+                                                        "Average",
+                                                        varKey);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
```
Comment (Myoldmopar) [inline src/EnergyPlus/DaylightingManager.cc:4635]: That output variable was not setup anywhere prior to this?  Either way this looks good.

File: src/EnergyPlus/DaylightingManager.cc [hunk 7]
```diff
@@ -5323,31 +5333,6 @@ namespace DaylightingManager {
...
-                        SetupOutputVariable("Daylighting Window Reference Point " + std::to_string(refPtNum) + " View Luminance",
-                                            OutputProcessor::Unit::cd_m2,
-                                            SurfaceWindow(SurfLoop).LumWinFromRefPtRep(refPtNum),
-                                            "Zone",
-                                            "Average",
-                                            Surface(SurfLoop).Name);
-                    }
-                }
-            }
-        }
```
Comment (Myoldmopar) [inline src/EnergyPlus/DaylightingManager.cc:5350]: OK there it is!

File: src/EnergyPlus/DaylightingManager.cc [hunk 8]
```diff
@@ -5615,48 +5600,21 @@ namespace DaylightingManager {
...
-                        ObjexxFCL::gio::write(OutputFileDebug, fmtA) << "     3,   !- Lighting Control Type";
-                        ObjexxFCL::gio::write(OutputFileDebug, fmtA)
-                            << "   0.0,   !- Glare Calculation Azimuth Angle of View Direction Clockwise from Zone y-Axis";
-                        ObjexxFCL::gio::write(OutputFileDebug, fmtA) << "      ,   !- Maximum Allowable Discomfort Glare Index";
-                        ObjexxFCL::gio::write(OutputFileDebug, fmtA) << "   0.0,   !- Minimum Input Power Fraction for Continuous Dimming Control";
-                        ObjexxFCL::gio::write(OutputFileDebug, fmtA) << "   0.0,   !- Minimum Light Output Fraction for Continuous Dimming Control";
-                        ObjexxFCL::gio::write(OutputFileDebug, fmtA) << "     0,   !- Number of Stepped Control Steps";
-                        ObjexxFCL::gio::write(OutputFileDebug, fmtA) << "   0.0;   !- Probability Lighting will be Reset When Needed in Manual Stepped Control";
-
-                        CheckTDDZone(Surface(SurfNum).Zone) = false;
```
Comment (Myoldmopar) [inline src/EnergyPlus/DaylightingManager.cc:5658]: All gone?  I'm sure I'll find this one shortly as well...

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 9]
```diff
@@ -1085,6 +1085,10 @@ namespace SurfaceGeometry {
 
         TotSurfaces = SurfNum + AddedSubSurfaces + NeedToAddSurfaces + NeedToAddSubSurfaces;
 
+        if (ErrorsFound) {
+            ShowFatalError(RoutineName + "Errors discovered, program terminates.");
+        }
```
Comment (Myoldmopar) [inline src/EnergyPlus/SurfaceGeometry.cc:1089]: This wasn't called before?  Yikes.

**Review by Myoldmopar — APPROVED** [review]
I will test this out locally but I have no issues with the code changes.

File: doc/engineering-reference/src/daylighting-and-window-calculations/daylighting-calculations.tex (hunk 1 above)
Comment (mjwitte) [reply doc/engineering-reference/src/daylighting-and-window-calculations/daylighting-calculations.tex:17]: Oops forgot about that. Feel free to change it here or I'll clean it up later.

File: src/EnergyPlus/DataSurfaces.hh (hunk 3 above)
Comment (mjwitte) [reply src/EnergyPlus/DataSurfaces.hh:1090]: It got left behind during the expansion of daylighting to allow more than one reference point. It wasn't being used anymore.

File: src/EnergyPlus/DataViewFactorInformation.hh (hunk 4 above)
Comment (mjwitte) [reply src/EnergyPlus/DataViewFactorInformation.hh:84]: Actually, not sure not that this is being used - I'll need to look at that later.

File: src/EnergyPlus/DaylightingManager.cc (hunk 5 above)
Comment (mjwitte) [reply src/EnergyPlus/DaylightingManager.cc:340]: Yep. By including all of the enclosure surfaces in the daylighting zone data, lots of other code just works as-is, because it's basically taking a reference point and looping through all the surfaces that it sees.  The next reference point may see many of the same surfaces, but that's ok, it sees all the ones it's supposed to. Not very different from two reference points in the same zone.

File: src/EnergyPlus/DaylightingManager.cc (hunk 8 above)
Comment (mjwitte) [reply src/EnergyPlus/DaylightingManager.cc:5658]: Yes, all gone. So, this was actually writing out a daylighting control object to the debug file and it never got updated to the newer syntax. I felt this was unnecessary.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 9 above)
Comment (mjwitte) [reply src/EnergyPlus/SurfaceGeometry.cc:1089]: It was, but much later. If I remember correctly I had to add this to prevent a zero subscript array bounds problem soon after this in new code that's asking for the enclosure number.

File: src/EnergyPlus/DataViewFactorInformation.hh (hunk 4 above)
Comment (mjwitte) [reply src/EnergyPlus/DataViewFactorInformation.hh:84]: Oh never mind, of course this is being used, quite bit.

**Comment by rraustad** [comment]
Apparently @mjwitte got in the "just, one, more, thing . . ." of interest so this is ready to merge?

**Comment by Myoldmopar** [comment]
Yeah, I think he felt it was fully ready to go.  I've looked it over and don't have any major issues but I will play around with it before merging.  Feel free to review/test as much as you want/can as well.

**Comment by rraustad** [comment]
CI results now show expected results. "quotes in LaTeX" still an issue?

**Comment by Myoldmopar** [comment]
Quotes issue is not a dealbreaker.  I haven't had time to play much with this but if you have then feel free to merge at will.  Otherwise I will soon.

**Comment by mjwitte** [comment]
@Myoldmopar Added code to disable the InteriorWindows option, throw a warning and proceed with the GroupedZones option for Solar and Daylighting.
